### PR TITLE
Add support for EE895, HTE501 and TEE501 sensors

### PR DESCRIPTION
**Pull request by Stock-M**
# What does this implement/fix?

Add support for EE895, HTE501 and TEE501 sensors.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes 

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2241

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:

```yaml

sensor:
  - platform: ee895
    co2:
      name: "Office CO2"
      accuracy_decimals: 0
    temperature:
      name: "Office Temperature"
      accuracy_decimals: 2
    pressure:
      name: "Office Pressure"
      accuracy_decimals: 1
    address: 0x5F
    update_interval: 15s

sensor:
  - platform: hte501
    temperature:
      name: "Office Temperature"
    humidity:
      name: "Office Humidity"
    address: 0x40
    update_interval: 60s

sensor:
  - platform: tee501
    name: "Office Temperature"
    address: 0x48
    update_interval: 60s

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by nielsnl68** [comment]
Hi Mathias, 
it is good that you work hard on this component set, but could you do me (us) a favor, and commit less frequent, my mailbox is filling up with all your commit messages. 

When you do this for testing, maybe you can create an external component and test it that way.

File: esphome/components/ee895/ee895.cpp [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    uint16_t crc16_check = 0;
+    uint8_t i2cResponse[8];
+    this->read(i2cResponse, 8);
+    crc16_check = (i2cResponse[7] << 8) + i2cResponse[6];
+    if (crc16_check != calcCrc16(i2cResponse, 7)) {
+      this->error_code_ = CRC_CHECK_FAILED;
+      this->status_set_warning();
+      return;
+    }
+    uint32_t x = i2cResponse[4] * 16777216 + i2cResponse[5] * 65536 + i2cResponse[2] * 256 + i2cResponse[3];
```
Comment (GregJArnold) [inline esphome/components/ee895/ee895.cpp]: Why did you do multiplication here?  Why not bit-shifting like everywhere else?  Also, I think you can directly store the data into a float like this:
```c
float temperature = (i2cResponse[4] << 24 | i2cResponse[5] << 16 | i2cResponse[2]  << 8 | i2cResponse[3]);

File: esphome/components/ee895/ee895.cpp [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+  LOG_UPDATE_INTERVAL(this);
+  LOG_SENSOR("  ", "Temperature", this->temperature_sensor_);
+  LOG_SENSOR("  ", "CO2", this->co2_sensor_);
+  LOG_SENSOR("  ", "Pressure", this->pressure_sensor_);
+}
+
+float EE895Component::get_setup_priority() const { return setup_priority::DATA; }
+
+void EE895Component::update() {
+  uint8_t address_1[] = {0x03, 0x03, 0xEA, 0x00, 0x02, 0xE8, 0xC5};
```
Comment (GregJArnold) [inline esphome/components/ee895/ee895.cpp]: This is a magic number.  You should create a function like `EE895Component::write_command(uint16_t addr, uint16_t reg_cnt)` that builds this, computes the checksum, and calls `this->write()`.

File: esphome/components/ee895/ee895.cpp [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    uint8_t i2cResponse[8];
+    this->read(i2cResponse, 8);
+    crc16_check = (i2cResponse[7] << 8) + i2cResponse[6];
+    if (crc16_check != calcCrc16(i2cResponse, 7)) {
+      this->error_code_ = CRC_CHECK_FAILED;
+      this->status_set_warning();
+      return;
+    }
+    uint32_t x = i2cResponse[4] * 16777216 + i2cResponse[5] * 65536 + i2cResponse[2] * 256 + i2cResponse[3];
+    float temperature = *(float *) &x;
```
Comment (GregJArnold) [inline esphome/components/ee895/ee895.cpp]: This code is common across all of your float reads.  You should create a `float EE895Component::read_float()` function that handles reading, computing the checksum, and squishing it into a float and returning it.

**Review by GregJArnold — CHANGES_REQUESTED** [review]
A couple refactor suggestions, for readability.

File: esphome/components/ee895/ee895.cpp [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+  uint32_t serial_number1 =
+      (serial_number[2] << 24) + (serial_number[3] << 16) + (serial_number[4] << 8) + serial_number[5];
+  uint32_t serial_number2 =
+      (serial_number[6] << 24) + (serial_number[7] << 16) + (serial_number[8] << 8) + serial_number[9];
+  uint32_t serial_number3 =
+      (serial_number[10] << 24) + (serial_number[11] << 16) + (serial_number[12] << 8) + serial_number[13];
+  uint32_t serial_number4 =
+      (serial_number[14] << 24) + (serial_number[15] << 16) + (serial_number[16] << 8) + serial_number[17];
+  ESP_LOGV(TAG, "    Serial Number: 0x%08X%08X%08X%08X", serial_number1, serial_number2, serial_number3,
+           serial_number4);
```
Comment (GregJArnold) [inline esphome/components/ee895/ee895.cpp]: I just found this out, but there's a `format_hex()` function in the core.
```suggestion
  ESP_LOGV(TAG, "    Serial Number: 0x%s", format_hex(serial_number + 2, 16).c_str());
```

File: esphome/components/hte501/hte501.cpp [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  if (identification[8] != calc_crc8_(identification, 0, 7)) {
+    this->error_code_ = CRC_CHECK_FAILED;
+    this->mark_failed();
+    return;
+  }
+  uint32_t serial_number1 =
+      (identification[0] << 24) + (identification[1] << 16) + (identification[2] << 8) + identification[3];
+  uint32_t serial_number2 =
+      (identification[4] << 24) + (identification[5] << 16) + (identification[6] << 8) + identification[7];
+  ESP_LOGV(TAG, "    Serial Number: 0x%08X%08X", serial_number1, serial_number2);
```
Comment (GregJArnold) [inline esphome/components/hte501/hte501.cpp]: Same as above.

File: esphome/components/tee501/tee501.cpp [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+  if (identification[8] != calc_crc8_(identification, 0, 7)) {
+    this->error_code_ = CRC_CHECK_FAILED;
+    this->mark_failed();
+    return;
+  }
+  uint32_t serial_number1 =
+      (identification[0] << 24) + (identification[1] << 16) + (identification[2] << 8) + identification[3];
+  uint32_t serial_number2 =
+      (identification[4] << 24) + (identification[5] << 16) + (identification[6] << 8) + identification[7];
+  ESP_LOGV(TAG, "    Serial Number: 0x%08X%08X", serial_number1, serial_number2);
```
Comment (GregJArnold) [inline esphome/components/tee501/tee501.cpp]: Same.

**Review by GregJArnold — CHANGES_REQUESTED** [review]

**Comment by jesserockz** [comment]
I'm gonna be a PITA and ask you to split this into 1 PR per integration please =)

**Comment by Stock-M** [comment]
I close this PR and open 3 new ones

**Pull request closed without merging**